text/0000-cloud-based-mapping.md

```diff
@@ -0,0 +1,134 @@
+
+| name                             | repo             | dependencies      |
+| -------------------------------- | ---------------- | ----------------- | 
+| libcartographer                  | cartographer     | proto3            |
+| cartographer_grpc                | cartographer     | proto3, gRPC      |
+| cartographer_(offline)_node      | cartographer_ros | proto3, ROS       |
+| cartographer_(offline)_node_grpc | cartographer_ros | proto3, gRPC, ROS |
+
+### Detailed Design
+
+#### cartographer_grpc
+
+The important interaction between local and global SLAM happens through `GlobalTrajectoryBuilder` between `LocalTrajectoryBuilder` and `PoseGraph`. 
+
+The following component diagram describes the cloud mapping architecture
+
+![component diagram](0000-assets/cloud-based-mapping-design.png "Component Diagram")
+
+On the agent side, a `GrpcMapBuilderBridge`is responsible for forwarding all information necessary for the formulation of the global SLAM problem in the cloud.
+This also includes the `LocalTrajectoryBuilder::InsertionResult`.
+
+Forwarding IMU, odometry and fixed frame data is straightforward:
+*sensor.proto* already contains proto messages for these data types (currently used only for file (de)serialization) and, as these flow through the `MapBuilderBridge` anyways, it is simple to forward them.
+
+To get access to the `LocalTrajectoryBuilder::InsertionResult`  the `GrpcMapBuilderBridge` will register a `LocalSlamResultCallback` with the `GlobalTrajectoryBuilder`.
+ 
+In order to properly sync submaps to the cloud, the `GrpcMapBuilderBridge` needs to be able to detect two types events:
+
+1. A new Submap has been created in `LocalTrajectoryBuilder`, that the `PoseGraph` does not yet know about.
+2. An existing Submap has been finished.
+
+i) can be detected by `GrpcMapBuilderBridge` remembering the last submap it has seen as `insertion_result.back()`.
+Whenever that changes between invocations of `AddRangeFinderData()`, a new submap has been added.
+`GrpcMapBuilderBridge` then sends up the `local_pose` of the new submap to the cloud.
+
+ii) can be detected by `GrpcMapBuilderBridge` observing `insertion_submaps.front().finished()`.
+Whenever that returns true (which is only once throughout the lifetime of any one `Submap`) the submap’s `Probability-` or `HybridGrid` is serialized and uploaded to the cloud.
+
+The `CollatedTrajectoryBuilder` feeds sensor data into the `GlobalTrajectoryBuilder` in a strict time order.
+This order needs to be maintained per sensor when sensor data is forwarded to the cloud.
+A simple way to achieve this is to use a streaming gRPC connection, which ensures that “each side will always get the other’s messages in the order they were written”.
+We choose to forward the sensor data in a streaming fashion rather than batching them up into a state that includes the finished submap for two reasons
+
+1. gRPC prefers small message sizes. Streaming does its best to ensure that.
```

Comment:
Mention somewhere that we ignore the intermittent connection dropouts that are likely to happen in real systems and address them later somehow.

Reply:
Done.